components/clientComponents/forms/Review/reviewUtils.ts

```diff
@@ -0,0 +1,208 @@
+  elements: ReviewElement[];
+};
+
+export type ReviewElement = {
+  title: string;
+  values: string | FileInputResponse | DateObject | ReviewElement[];
+  element: FormElement | undefined;
+};
+
+export function formatElementValues(element: ReviewElement) {
+  if (!element.values) {
+    return "-";
+  }
+  // Case of a File upload
+  if ((element.values as FileInputResponse).based64EncodedFile !== undefined) {
+    const file = element.values as FileInputResponse;
+    if (!file.name || !file.size || file.size < 0) {
+      return "-";
+    }
+    const fileSizeInMB = (file.size / 1024 / 1024).toFixed(2);
+    return `${file.name} (${fileSizeInMB} MB)`;
+  }
+
+  // Case of a Formatted date
+  if (element.element?.type === FormElementTypes.formattedDate) {
+    return getFormattedDateFromObject(
+      element.element?.properties.dateFormat as DateFormat,
+      JSON.parse(element.values as string) as DateObject
+    );
+  }
+
+  // Case of Address Complete
+  if (element.element?.type === FormElementTypes.addressComplete) {
+    if (element.element.properties?.addressComponents?.splitAddress === true) {
+      return element.values as string; // We're a split address, broken into components.
+    } else {
+      return getAddressAsString(JSON.parse(element.values as string) as AddressElements);
+    }
+  }
+
+  // Case of an array like element e.g. checkbox
+  if (Array.isArray(element.values)) {
+    return element.values.join(", ") || "-";
+  }
+  // Case of a single value element e.g. input
+  return String(element.values);
+}
+
+export function getReviewItemElements(
+  groupElements: string[],
+  formElements: FormElement[],
+  matchedIds: string[],
+  formValues: FormValues,
+  lang: string,
+  addressCompleteStrings: AddressCompleteLabels
+) {
+  const shownFormElements = filterShownElements(formElements, matchedIds);
+  const shownElementIds = getElementIdsAsNumber(
+    filterValuesForShownElements(groupElements, shownFormElements)
+  );
+  return shownElementIds.map((elementId) => {
+    const element = formElements.find((item) => item.id === elementId);
+    const reviewElement: ReviewElement = {
+      title: element?.properties?.[getLocalizedProperty("title", lang)] as string,
+      values: formValues[elementId as unknown as keyof typeof formValues] as string,
+      element,
+    };
+
+    let resultValues: string | ReviewElement[] = formatElementValues(reviewElement);
+
+    // Handle any Sub Elements. Note Sub Elements = Dynamic Rows = Repeating Sets
+    if (element?.type === FormElementTypes.dynamicRow) {
+      resultValues = [];
+      const parentId = element.id;
+      const parentTitle = element.properties?.[getLocalizedProperty("title", lang)];
+      const subElements = element.properties?.subElements;
+      // Use FormValues as the source of truth and for each FormValue value, map the related
+      // subElement title to the FormValue value
+      const subElementValues = (formValues[parentId] as string[]).map(
```

Comment:
Wondering if we can break this up ---

We have 2 maps inside of a map

Reply:
Peter decides to add a 3 map in the maps because he likes maps in maps. (just kidding :)